### PR TITLE
DBZ-3124 Add ENUM column type support for Vitess 9.0.0

### PR DESCRIPTION
**Pull request by keweishang**
https://issues.redhat.com/browse/DBZ-3124

Related PRs:

- Website: https://github.com/debezium/debezium.github.io/pull/641
- Vitess Connector documentation: https://github.com/debezium/debezium/pull/2142

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -18,9 +18,9 @@
         <version.debezium>${project.version}</version.debezium>
 
         <!-- Dependencies -->
-        <version.vitess.grpc>7.0.0</version.vitess.grpc>
+        <version.vitess.grpc>9.0.0</version.vitess.grpc>
```
Comment (keweishang) [inline pom.xml:21]: I'll update the compatibility matrix on the Debezium website as well

Comment (keweishang) [reply pom.xml:21]: Corresponding compatibility matrix PR: https://github.com/debezium/debezium.github.io/pull/641

File: src/main/java/io/debezium/connector/vitess/VitessType.java [hunk 2]
```diff
@@ -14,10 +19,17 @@
...
+    private final List<String> enumValues;
 
     public VitessType(String name, int jdbcId) {
+        this(name, jdbcId, null);
+    }
+
+    public VitessType(String name, int jdbcId, List<String> enumValues) {
         this.name = name;
         this.jdbcId = jdbcId;
+        this.enumValues = enumValues;
```
Comment (gunnarmorling) [inline src/main/java/io/debezium/connector/vitess/VitessType.java:32]: Can we wrap this into Collections.unmodifiableList(enumValues) here?

File: src/main/java/io/debezium/connector/vitess/VitessType.java [hunk 3]
```diff
@@ -14,10 +19,17 @@
     private final String name;
     // enum of column jdbc type
     private final int jdbcId;
+    // permitted enum values
+    private final List<String> enumValues;
 
     public VitessType(String name, int jdbcId) {
+        this(name, jdbcId, null);
```
Comment (gunnarmorling) [inline src/main/java/io/debezium/connector/vitess/VitessType.java:26]: Let's use Collections.emptyList() instead.

File: src/main/java/io/debezium/connector/vitess/VitessType.java [hunk 4]
```diff
@@ -28,27 +40,59 @@ public int getJdbcId() {
         return jdbcId;
     }
 
+    public List<String> getEnumValues() {
+        return enumValues;
+    }
+
+    public boolean isEnum() {
+        return enumValues != null && enumValues.size() != 0;
```
Comment (gunnarmorling) [inline src/main/java/io/debezium/connector/vitess/VitessType.java:48]: As per above, you then can simply do return !enumValues().isEmpty().

File: src/main/java/io/debezium/connector/vitess/VitessValueConverter.java [hunk 5]
```diff
@@ -34,9 +37,13 @@ public VitessValueConverter(
     @Override
     public SchemaBuilder schemaBuilder(Column column) {
         String typeName = column.typeName().toUpperCase();
-        if (matches(typeName, "JSON")) {
+        if (matches(typeName, Query.Type.JSON.name())) {
             return Json.builder();
         }
+        if (matches(typeName, Query.Type.ENUM.name())) {
+            String commaSeperatedOptions = Strings.join(",", column.enumValues());
```
Comment (gunnarmorling) [inline src/main/java/io/debezium/connector/vitess/VitessValueConverter.java:44]:
```suggestion
            String commaSeparatedOptions = Strings.join(",", column.enumValues());
```

File: src/main/java/io/debezium/connector/vitess/VitessValueConverter.java [hunk 6]
```diff
@@ -34,9 +37,13 @@ public VitessValueConverter(
...
+            return io.debezium.data.Enum.builder(commaSeperatedOptions);
```
Comment (gunnarmorling) [inline src/main/java/io/debezium/connector/vitess/VitessValueConverter.java:45]: How about we add an overload to `build()` which takes the values, making usage for callers like this a bit simpler?

**Review by gunnarmorling — COMMENTED** [review]
LGTM; some small suggestions inline.

File: src/main/java/io/debezium/connector/vitess/VitessType.java (hunk 2 above)
Comment (keweishang) [reply src/main/java/io/debezium/connector/vitess/VitessType.java:32]: Good point. Addressed

File: src/main/java/io/debezium/connector/vitess/VitessType.java (hunk 3 above)
Comment (keweishang) [reply src/main/java/io/debezium/connector/vitess/VitessType.java:26]: Good point. Addressed

File: src/main/java/io/debezium/connector/vitess/VitessType.java (hunk 4 above)
Comment (keweishang) [reply src/main/java/io/debezium/connector/vitess/VitessType.java:48]: Good point. Addressed

File: src/main/java/io/debezium/connector/vitess/VitessValueConverter.java (hunk 5 above)
Comment (keweishang) [reply src/main/java/io/debezium/connector/vitess/VitessValueConverter.java:44]: fixed

File: src/main/java/io/debezium/connector/vitess/VitessValueConverter.java (hunk 6 above)
Comment (keweishang) [reply src/main/java/io/debezium/connector/vitess/VitessValueConverter.java:45]: Of course. I'll create a PR to overload the `builder()` static function in the Debezium main repo.

Comment (keweishang) [reply src/main/java/io/debezium/connector/vitess/VitessValueConverter.java:45]: I added the `builder()` overload to the existing Debezium main repo [PR](https://github.com/debezium/debezium/pull/2142). Would that be ok? I could also create a separate ticket and PR.

Comment (keweishang) [reply src/main/java/io/debezium/connector/vitess/VitessValueConverter.java]: I'll wait for the Debezium PR to be merged and built use the overloaded `builder()` here.

Comment (gunnarmorling) [reply src/main/java/io/debezium/connector/vitess/VitessValueConverter.java]: Just merged the core PR :)

**Comment by gunnarmorling** [comment]
@ani-sha, just seeing that the JIRA-key-less commit message didn't trigger a build failure here. Could you look into setting up the commit message job for this (and potentially other) repo? Thx!

**Comment by gunnarmorling** [comment]
@keweishang, sorry, just noticed one more thing is missing for the docs update in core: the data type table still lists enums as unsupported.

**Comment by keweishang** [comment]
> just noticed one more thing is missing for the docs update in core: the data type table still lists enums as unsupported.

@gunnarmorling, sorry I missed that. I just created a PR for updating the documentation - data type table lists enums as supported:
https://github.com/debezium/debezium/pull/2149

**Comment by keweishang** [comment]
@ani-sha I'll push an amend-commit (a.k.a. `git commit --amend`) to modify the commit message, so that every commit would have the issue key. But here's the screenshot before I pushed my amend-commit:
image

**Review by gunnarmorling — APPROVED** [review]
LGTM now, thx! Can you merge yourself and resolve the issue once CI has passed? Thanks a lot!

**Comment by keweishang** [comment]
Sure, I'll merge the PR and resolve the issue once CI has passed.

**Pull request merged**